packages/opencensus-exporter-trace-stackdriver/test/test-stackdriver.ts

```diff
@@ -0,0 +1,203 @@
+        assert.strictEqual(
+            exporter.exporterBuffer.getQueue()[0].spans[0].name, spanName);
+      });
+    });
+  });
+
+  /* Should export spans to stackdriver */
+  describe('publish()', () => {
+    it('should fail exporting by authentication error', () => {
+      process.env.GOOGLE_APPLICATION_CREDENTIALS = '';
+      if (dryrun) {
+        nocks.oauth2((body) => true);
+      }
+
+      return tracer.startRootSpan(
+          {name: 'sdNoExportTestRootSpan'}, async (rootSpan) => {
+            const span = tracer.startChildSpan('sdNoExportTestChildSpan');
+            span.end();
+            rootSpan.end();
+
+            return exporter.publish([rootSpan]).then((result: string) => {
+              assert.ok(result.indexOf('authorize error') >= 0);
+              assert.strictEqual(
+                  exporter.failBuffer[0].traceId, rootSpan.spanContext.traceId);
+            });
+          });
+    });
+
+    it('should fail exporting with wrong projectId', () => {
+      const NOEXIST_PROJECT_ID = 'no-existent-project-id-99999';
+      if (dryrun) {
+        process.env.GOOGLE_APPLICATION_CREDENTIALS =
+            __dirname + '/fixtures/fakecredentials.json';
+        nocks.oauth2((body) => true);
+        const scope = nocks.patchTraces(NOEXIST_PROJECT_ID, null, null, false);
+      }
+      const failExporterOptions = {
+        projectId: NOEXIST_PROJECT_ID,
+        logger: logger.logger('debug')
+      };
+      const failExporter = new StackdriverTraceExporter(failExporterOptions);
+      const failTracer = new classes.Tracer();
+      failTracer.start({samplingRate: 1});
+      failTracer.registerEndSpanListener(failExporter);
+      return failTracer.startRootSpan(
+          {name: 'sdNoExportTestRootSpan'}, async (rootSpan) => {
+            const span = failTracer.startChildSpan('sdNoExportTestChildSpan');
+            span.end();
+            rootSpan.end();
+
+            return failExporter.publish([rootSpan]).then((result: string) => {
+              assert.ok(result.indexOf('sendTrace error: ') >= 0);
+
+              assert.strictEqual(
+                  failExporter.failBuffer[0].traceId,
+                  rootSpan.spanContext.traceId);
+            });
+          });
+    });
+
+    it('should export traces to stackdriver', () => {
+      if (dryrun) {
+        nocks.oauth2((body) => true);
+        const scope = nocks.patchTraces(PROJECT_ID, null, null, false);
+      }
+
+      return tracer.startRootSpan(
+          {name: 'sdExportTestRootSpan'}, async (rootSpan) => {
+            const span = tracer.startChildSpan('sdExportTestChildSpan');
+            span.end();
+            rootSpan.end();
+
+            return exporter.publish([rootSpan]).then((result: string) => {
+              assert.ok(result.indexOf('sendTrace sucessfully') >= 0);
+            });
+          });
+    });
+
+
+    it('should fail exporting by network error', async () => {
```

Comment:
If `dryrun` is set to false, do we expect a real network error to occur here?

Reply:
Hmm... in that case is it necessary to have network access at all for this test? We don't need to authenticate with a real server if we're just going to hit a fake endpoint that will fail.

Reply:
In fact it's not necessary to authenticate with a real server. Changed to always authenticate with nock.